File: lib/src/determinism/dom_stabilization.dart
import '../crypto/hashing.dart' show computeDeterministicHash;

final _uuidRe = RegExp(
  r'[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}',
  caseSensitive: false,
);
final _timestampRe = RegExp(
  r'\b\d{4}-\d{2}-\d{2}[T ]\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:Z|[+-]\d{2}:?\d{2})?\b',
);
final _reactRe = RegExp(r'data-react[a-z0-9-]*="[^"]*"', caseSensitive: false);
final _vueRe = RegExp(r'data-v-[a-z0-9]+="[^"]*"', caseSensitive: false);
final _angularRe = RegExp(
  r'\s(?:ng-version|_ngcontent[a-z0-9-]*|_nghost[a-z0-9-]*)="[^"]*"',
  caseSensitive: false,
);
final _nonceRe = RegExp(r'\snonce="[^"]*"', caseSensitive: false);
final _scriptBodyRe =
    RegExp(r'<script\b[^>]*>[\s\S]*?</script>', caseSensitive: false);

String stabilizeDomHtml(String html) {
  var s = html;
  s = s.replaceAll(_uuidRe, 'uuid-stabilized');
  s = s.replaceAll(_timestampRe, 'timestamp-stabilized');
  s = s.replaceAll(_reactRe, '');
  s = s.replaceAll(_vueRe, '');
  s = s.replaceAll(_angularRe, '');
  s = s.replaceAll(_nonceRe, '');
  s = s.replaceAll(_scriptBodyRe, '<script>stabilized</script>');
  s = s.replaceAll(RegExp(r'<!--[\s\S]*?-->'), '');
  s = s.replaceAll(RegExp(r'\s+'), ' ').trim();
  return s;
}

String computeStableDomHash(String html) =>
    computeDeterministicHash(stabilizeDomHtml(html));

String computeSpaFingerprint(String html) => computeStableDomHash(html);

File: src/determinism/domStabilization.ts
import { computeDeterministicHash } from "../crypto/kaalkaRuntime.js";

const UUID_RE =
  /[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}/gi;
const TIMESTAMP_RE =
  /\b\d{4}-\d{2}-\d{2}[T ]\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:Z|[+-]\d{2}:?\d{2})?\b/g;
const REACT_RE = /data-react[a-z0-9-]*="[^"]*"/gi;
const VUE_RE = /data-v-[a-z0-9]+="[^"]*"/gi;
const ANGULAR_RE = /\s(?:ng-version|_ngcontent[a-z0-9-]*|_nghost[a-z0-9-]*)="[^"]*"/gi;
const NONCE_RE = /\snonce="[^"]*"/gi;
const SCRIPT_BODY_RE = /<script\b[^>]*>[\s\S]*?<\/script>/gi;

/** Production DOM stabilization for deterministic replay fingerprints. */
export function stabilizeDomHtml(html: string): string {
  let s = html;
  s = s.replace(UUID_RE, "uuid-stabilized");
  s = s.replace(TIMESTAMP_RE, "timestamp-stabilized");
  s = s.replace(REACT_RE, "");
  s = s.replace(VUE_RE, "");
  s = s.replace(ANGULAR_RE, "");
  s = s.replace(NONCE_RE, "");
  s = s.replace(SCRIPT_BODY_RE, "<script>stabilized</script>");
  s = s.replace(/<!--[\s\S]*?-->/g, "");
  s = s.replace(/\s+/g, " ").trim();
  return s;
}

export function normalizeDom(html: string): string {
  return stabilizeDomHtml(html);
}

export function computeStableDomHash(html: string): string {
  return computeDeterministicHash(stabilizeDomHtml(html));
}

/** SPA fingerprint: hash of stabilized DOM (not raw HTML). */
export function computeSpaFingerprint(html: string): string {
  return computeStableDomHash(html);
}
